### ISSUE TITLE
tensor types.

### ISSUE THREAD

**Issue by nicholas-leonard** [nicholas-leonard/dp#20]
It is really hard to abstract away the differences in cuda and non-cuda tensors. Ideally, the user need only say experiment:cuda() and/or datasource:cuda(), and we use cuda as much as possible. 

Obstacles:
- Not all `nn.modules` are implemented in cunn
- Not all `torch.Tensor` methods and functions are implemented in cutorch.

If I have a Concat container which takes as input a single datatensor and broadcasts its modules, some might require a cuda tensor, others a float tensor. How can I solve this problem efficiently and easily at the same time?
1. Currently, we leave it to _forward and _backward methods to specify the tensor views (`feature()`, `image()`, etc). 
   - We could just add a parameter to these views which asks for the type of the returned tensor. 
   - We could automate this with an activation() method, which could use the self._tensor_type as default. 
   - Subclasses that have different input views could hard-code something else.
2. We could also force the tensors returned by `BaseTensors` to be read-only.
   - They are only used for activations and their gradients, which are written once.  
   - Unless of course we decide to reuse them from batch to batch...
   - But this would be a bad idea and would lead to little or no speedup since `nn.Module` and `dp.Batch` already reuse the torch.Tensors from batch to batch.
   - So then, this way, we can cache different tensor types such that a different modules can call different `BaseTensor` views and reuse work done by previous ones?
   - Yup.
